Stage_10

KinCadeFoster opened the pull request:
- доработка бота

KinCadeFoster committed:
Stage_10
- доработка бота

KinCadeFoster committed:
Stage_10
- обновил список слов на изучение

danelyan commented on src/main/kotlin/Telegram.kt:
```diff
@@ -1,83 +1,124 @@
+    trainer: LearnWordsTrainer,
+    chatId: Long,
+    telegramBotService: TelegramBotService,
+    json: Json
```
Обрати внимание, Json можно создать внутри TelegramBotService, и не пробрасывать его извне каждый раз

KinCadeFoster replied:
done

danelyan commented on src/main/kotlin/Telegram.kt:
```diff
@@ -1,83 +1,124 @@
-                }
-            }
+    when {
+        text?.lowercase() == "/start" -> {
```
`/start` лучше унести в константы

KinCadeFoster replied:
done

danelyan reviewed (changes requested):
Круто! Почти все учтено ))

Три момента:

1. Как остановить процесс изучения слов? Давай выход в меню добавим последним пунктом
2. Давай разместим в столбик, больше влезет
3. Давай уберем кавычки из статистики:

> 
> "Выучено 0 из 900 слов. 
>  Это 0% от всех слов в базе данных!"


<img width="446" alt="Screenshot 2024-07-12 at 15 11 51" src="https://github.com/user-attachments/assets/d2954eec-bb70-43bc-99a1-e2b834c40785">

KinCadeFoster committed:
Stage_10
- /start вынес в константы
- добавил кнопку для возсравта в главное меню при изучении слов
- добавил автоматический вызов telegramBotService.sendMenu для методов RESET_CLICKED и STATISTICS_CLICKED, что бы после выполнения мы оказывались сразу в меню
- изменил формирование кнопок, теперь двк строки по 2 ответа + кнопка возврата
- убрал лишние ковычки из текста статистики

KinCadeFoster committed:
Stage_10
- Json создаётся внутри TelegramBotService

danelyan commented on src/main/kotlin/TelegramBotService.kt:
```diff
@@ -1,135 +1,152 @@
+            )
+            if (row.size == 2) {
+                inlineKeyboard.add(row.toList())
+                row.clear()
```
круто! Вот тебе еще альтернативный вариант с использованием chunked(N). Эта функция разбивает список на подсписки в которых N элементов:
```Kotlin
       inlineKeyboard.addAll(
            variants.chunked(2).mapIndexed { rowIndex, words: List<Word> ->
                words.mapIndexed { colIndex, word ->
                    println(rowIndex * 2 + colIndex)
                }
            }
        )
```

danelyan commented on src/main/kotlin/TelegramBotService.kt:
```diff
@@ -1,135 +1,152 @@
+            .header("Content-type", "application/json")
+            .POST(HttpRequest.BodyPublishers.ofString(requestBodyString))
+            .build()
+        val response: HttpResponse<String> = client.send(request, HttpResponse.BodyHandlers.ofString())
```
Вот эти места (выполнение сетевых запросов при помощи `send()`) - не защищены, поэтому любая сетевая проблема приведет к падению приложения. Попробуй выключить интернет ненадолго, запустив бота локально.

Точно также может стать недоспуным сервер телеграма.
Можно сделать просто обертку типа: `runCatching  { client.send(...) }`

И пропускать итерацию цикла, если есть запрос выполнен не успешно

KinCadeFoster committed:
Stage_10
- добавилд обработку сетевых вызовов с помощью runCatching

danelyan reviewed (approved):
![yougotthis-martini](https://github.com/Gaileks/LearnEnglishWordsTelegramBot/assets/7125622/3339b61b-9b81-484f-989b-a6e73733be08)

Поздравляю! Ты сделал бота на Kotlin.

Основная часть работы пройдена. Теперь этот бот твой - можешь развивать его как хочешь. Дальше я подготовил тебе развилку из 6 простеньких доработок, все они уже в боте. Однако можешь не ограничиваться только ими, любые идеи приветствуются, просто опиши в личку и мы продумаем детали. 

Дальше как обычно - создаешь Pull Request (даже если что-то еще не работает - шлешь), а я пишу что можно улучшить.
